File: lib/tmdb.js
var request = require('request');
var Promise = require('bluebird');
var config = require('./config');
var _ = require('lodash');
var moment = require('moment');

// request.debug = true;

var makeRequest = function(uri, data) {

  var opts = {
    url: "http://api.themoviedb.org/3" + uri,
    method: 'GET',
    json: true
  }

  opts.qs = _.defaults(data || {}, {
    api_key: config.TMDB_KEY
  });

  return Promise.resolve()
    .then(function() {
      return new Promise(function(resolve, reject) {
        request(opts, function(err, response, body) {
          if (err) return reject(err);
          return resolve([response, body]);
        })
      })
    })
    .spread(function(response, body) {
      if (response.statusCode == 429) {
        return Promise.resolve()
          .delay(response.headers['retry-after'] * 1100)
          .then(function() {
            return makeRequest(uri, data)
          })
      }

      if (response.statusCode !== 200) {
        throw new Error("TMDB responded with " + response.statusCode + " instead of 200")
      }

      return body
    })

}

var formatMovie = function(movie) {
  var details =  _.pick(movie, ["title", "release_date", "popularity", "vote_average", "id", "vote_count", "poster_path"])
  details.poster_url = config.TMDB_POSTER_URL + details.poster_path;
  return details;
}

var getMovies = function(page) {

  return makeRequest('/discover/movie', {
    'sort_by': 'popularity.desc',
    'vote_count.gte': 25,
    'vote_average.gte': 4,
    'page': page || 1,
    'language': 'en',
    'release_date.gte': moment().subtract(1.5, 'year').format('YYYY-MM-DD'),
    'release_date.lte': moment().subtract(30, 'days').format('YYYY-MM-DD')
  })

}

module.exports.getMovies = function() {

  return Promise.resolve(getMovies())
    .bind({
      movies: []
    })
    .then(function(response) {

      var movies = this.movies;

      _.each(response.results, function(result) {
        movies.push(result)
      })

      return _.range(response.page, response.total_pages)
    })
    .map(function(page) {

      return getMovies(page)
        .then(function(response) {
          return response.results
        })

    }, {concurrency: 1})
    .each(function(results) {

      var movies = this.movies;

      _.each(results, function(result) {
        movies.push(result);
      })

    })
    .then(function() {

      this.movies = _.compact(this.movies);
      return this.movies

    })
    .map(function(movie) {
      return formatMovie(movie)
    })

}

module.exports.getMovie = function(id) {

  return makeRequest('/movie/' + id)

}

module.exports.searchMovie = function(title, year) {

  var query = {
    query: title.replace(/[^\w\s]/gi, '')
  }

  if (year) {
    query.year = year
  }

  return makeRequest('/search/movie', query)
    .then(function(results) {
      return _.first(results.results)
    })
    .then(function(movie) {
      return formatMovie(movie)
    })

}

File: app/lib/tmdb.py
import requests
import json
import time
from urllib.parse import urlencode
from datetime import datetime, timedelta, date

def call_tmdb_api(apikey,uri):
    retry = True
    while retry:
        Authorization = "Bearer " + apikey
        headers = { 
            "Authorization": Authorization,
            "Content-Type": "application/json;charset=utf-8"
        }
        url = "http://api.themoviedb.org/3/{}".format(uri)
        response = requests.get(url, headers=headers)
        if response.status_code == 429:
            sleeptime = response.headers['retry-after'] * 1100
            return
        if response.status_code != 200:
            print("Warning: TMDB request failed: {}".format(response.status_code))
            return response
        output = json.loads(response.content.decode('utf-8'))
        retry = False
    return output

def find_tmdb_movie_by_imdbid(apikey,imdbId):
    uri = "find/{}?api_key={}&language=en-US&external_source=imdb_id".format(imdbId,apikey)            
    info = call_tmdb_api(apikey,uri)
    return info

def get_tmdb_popular_movie_query_string(page=1):
    year = timedelta(days=365)
    parms = {
        'sort_by': 'popularity.desc',
        'vote_count.gte': 25,
        'vote_average.gte': 4,
        'page': page,
        'language': 'en',
        'release_date.gte': (date.today()-(year*1.5)).isoformat(),
        'release_date.lte': (date.today()-(timedelta(days=30))).isoformat()
    }
    output = urlencode(parms)
    return output

def get_tmdb_popular_movies(apikey):
    page = 0
    total_pages = 1
    results = []
    while page <= total_pages:
        page += 1
        query_string = get_tmdb_popular_movie_query_string()
        uri = "discover/movie?api_key={}&{}".format(apikey,query_string)
        result = call_tmdb_api(apikey,uri)
        results += result['results']
        page = result['page']
        total_pages = result['total_pages']
    return results


def get_tmdb_movie(apikey,tmdbid):
    uri = "movie/{}?api_key={}".format(tmdbid, apikey)
    response = call_tmdb_api(apikey,uri)
    return response

def search_tmdb_movie(apikey,title,year):
    query = {
        "query": title.replace("/[^\w\s]/gi", ''),
        "year": year
    }
    query_string = urlencode(query)    
    uri = "search/movie?api_key={}&{}".format(apikey,query_string)
    response = call_tmdb_api(apikey,uri)
    return response
